src/rgw/rgw_coroutine.cc

```diff
@@ -143,22 +143,30 @@ RGWCoroutine::~RGWCoroutine() {
 
 void RGWCoroutine::init_new_io(RGWIOProvider *io_provider)
 {
+  ceph_assert(stack); // if there's no stack, io_provider won't be uninitialized
   stack->init_new_io(io_provider);
 }
 
 void RGWCoroutine::set_io_blocked(bool flag) {
-  stack->set_io_blocked(flag);
+  if (stack) {
+    stack->set_io_blocked(flag);
+  }
 }
 
 void RGWCoroutine::set_sleeping(bool flag) {
-  stack->set_sleeping(flag);
+  if (stack) {
+    stack->set_sleeping(flag);
+  }
 }
 
 int RGWCoroutine::io_block(int ret, int64_t io_id) {
   return io_block(ret, rgw_io_id{io_id, -1});
 }
 
 int RGWCoroutine::io_block(int ret, const rgw_io_id& io_id) {
+  if (!stack) {
+    return 0; // XXX: what does 0 mean?
```

Comment:
Does the returned value matters here anyway? (we should change the comment)
Maybe we should add more checks as you've mentioned to the code where the stack is being called.
Other than those, LGTM

Reply:
> Does the returned value matters here anyway?

i see that several coroutines in rgw_sync.cc and rgw_data_sync.cc are returning the result with `return io_block(0);`

if coroutines return an error without calling `set_cr_done()` or `set_cr_error()`, it looks like we'd trigger this assertion in [RGWCoroutinesStack::operate()](https://github.com/ceph/ceph/blob/master/src/rgw/rgw_coroutine.cc#L246-L266):
```c++
/* should r ever be negative at this point? */
ceph_assert(r >= 0);
```
but if we're in `RGWCoroutinesStack::operate()`, then `RGWCoroutine::stack` probably isn't null, so maybe it doesn't matter in this specific case

> (we should change the comment)

i removed the comment, thanks

> Maybe we should add more checks as you've mentioned to the code where the stack is being called

the crashes do seem to be specific to `RGWOmapAppend`, so we could resolve https://tracker.ceph.com/issues/49302 by changing its callers to hold an extra reference to its stack. but who knows where else this is broken?

i'm hoping that this change in `RGWCoroutine` fixes this entire class of bugs, without needing to make the calling code more complicated. but this coroutine stuff itself is complicated, so we may just be breaking other things